src/main/scala/com/exasol/cloudetl/kinesis/KinesisImportQueryGenerator.scala

```diff
@@ -0,0 +1,53 @@
+package com.exasol.cloudetl.kinesis
+
+import scala.collection.JavaConverters._
+
+import com.exasol.{ExaImportSpecification, ExaMetadata}
+import com.exasol.cloudetl.kinesis.KinesisConstants._
+
+/**
+ * This object returns an SQL query for an EXASOL IMPORT FROM SCRIPT statement.
+ */
+object KinesisImportQueryGenerator {
+
+  /** Provides a SELECT query for IMPORT.
+   *
+   * This is a function which is called inside Exasol when an IMPORT query refers to a script
+   * based on this object.
+   */
+  def generateSqlForImportSpec(
+    exaMetadata: ExaMetadata,
+    importSpecification: ExaImportSpecification
+  ): String = {
+    val kinesisUserProperties = new KinesisUserProperties(
+      importSpecification.getParameters.asScala.toMap
+    )
+    val tableName = kinesisUserProperties.getTableName
+    val innerUdfProperties =
+      kinesisUserProperties.createSelectedPropertiesMap(
+        KinesisUserProperties.AWS_ACCESS_KEY_PROPERTY,
+        KinesisUserProperties.AWS_SECRET_KEY_PROPERTY,
+        KinesisUserProperties.AWS_SESSION_TOKEN_PROPERTY,
+        KinesisUserProperties.REGION_PROPERTY,
+        KinesisUserProperties.STREAM_NAME_PROPERTY
+      )
+    val innerUdfPropertiesString = kinesisUserProperties.makeString(innerUdfProperties)
+    s"""SELECT KINESIS_IMPORT(
+       |  '$innerUdfPropertiesString',
+       |  $KINESIS_SHARD_ID_COLUMN_NAME,
+       |  $KINESIS_SEQUENCE_NUMBER_COLUMN_NAME
+       |)
+       |FROM (
+       |  SELECT KINESIS_SHARDS.$KINESIS_SHARD_ID_COLUMN_NAME,
+       |  MAX($KINESIS_SEQUENCE_NUMBER_COLUMN_NAME) AS $KINESIS_SEQUENCE_NUMBER_COLUMN_NAME
+       |  FROM (
+       |    SELECT KINESIS_METADATA('$innerUdfPropertiesString')) AS KINESIS_SHARDS
+       |    LEFT JOIN $tableName
```

Comment:
The database join seems very expensive here, use the shard, max(sequence) result and join inside the metadata udf.

Reply:
done